File: src/Rust/tests/pipeline_test.rs
use std::io::{Read, Write};
use std::net::{TcpListener, TcpStream};
use std::panic::{catch_unwind, AssertUnwindSafe};
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::{mpsc, Arc, Barrier, Mutex};
use std::thread;
use std::time::Duration;

use pipeline_services::remote::http_step::{http_step, RemoteSpec};
use pipeline_services::{
  short_circuit, ActionControl, Pipeline, PipelineJsonLoader, PipelineObserver, PipelineProvider,
  PipelineProviderMode, PipelineRegistry, PipelineRouter,
};

const REMOTE_FIXTURE_BODY: &str = "Hello from remote fixture\n";

fn identity_action(value: String) -> String {
  value
}

fn append_and_stop(value: String) -> String {
  short_circuit();
  format!("{value}S")
}

fn failing_action(value: String) -> String {
  drop(value);
  panic!("boom")
}

#[derive(Clone, Debug, PartialEq, Eq)]
struct ImmutableContext {
  value: String,
  error: String,
}

#[test]
fn run_returns_context_and_detailed_uses_same_semantics() {
  let mut pipeline = Pipeline::new("simple", true);
  pipeline.add_pre_action(|value: String| format!("{value}P"));
  pipeline.add_action(|value: String| format!("{value}A"));
  pipeline.add_post_action(|value: String| format!("{value}Z"));

  assert_eq!(pipeline.run("X".to_string()), "XPAZ");
  let result = pipeline.run_detailed("X".to_string());
  assert_eq!(result.context, "XPAZ");
  assert_eq!(result.action_timings.len(), 3);
}

#[test]
fn short_circuit_preserves_phase_rules() {
  let mut main = Pipeline::new("main", true);
  main.add_action(append_and_stop);
  main.add_action(|value: String| format!("{value}B"));
  main.add_post_action(|value: String| format!("{value}P"));
  assert_eq!(main.run("X".to_string()), "XSP");

  let mut pre = Pipeline::new("pre", true);
  pre.add_pre_action(append_and_stop);
  pre.add_pre_action(|value: String| format!("{value}2"));
  pre.add_action(|value: String| format!("{value}M"));
  pre.add_post_action(|value: String| format!("{value}P"));
  assert_eq!(pre.run("X".to_string()), "XS2P");

  let mut post = Pipeline::new("post", true);
  post.add_action(|value: String| format!("{value}M"));
  post.add_post_action(append_and_stop);
  post.add_post_action(|value: String| format!("{value}2"));
  let result = post.run_detailed("X".to_string());
  assert_eq!(result.context, "XMS2");
  assert!(result.short_circuited);
}

#[test]
fn exception_policies_capture_errors_and_run_post_actions() {
  let mut continuing = Pipeline::new("continue", false);
  continuing.add_action(|value: String| format!("{value}A"));
  continuing.add_action(failing_action);
  continuing.add_action(|value: String| format!("{value}B"));
  let result = continuing.run_detailed("X".to_string());
  assert_eq!(result.context, "XAB");
  assert!(!result.short_circuited);
  assert_eq!(result.errors.len(), 1);
  assert_eq!(result.errors[0].action_index, 1);

  let mut stopping = Pipeline::new("stop", true);
  stopping.add_action(|value: String| format!("{value}A"));
  stopping.add_action(failing_action);
  stopping.add_action(|value: String| format!("{value}B"));
  stopping.add_post_action(|value: String| format!("{value}P"));
  let result = stopping.run_detailed("X".to_string());
  assert_eq!(result.context, "XAP");
  assert!(result.short_circuited);
}

#[test]
fn error_handler_updates_immutable_context() {
  let mut pipeline = Pipeline::new("immutable", false);
  pipeline.on_error_handler(|context: ImmutableContext, error| ImmutableContext {
    value: context.value,
    error: error.message,
  });
  pipeline.add_action(|context: ImmutableContext| {
    drop(context);
    panic!("bad")
  });
  pipeline.add_action(|mut context: ImmutableContext| {
    context.value.push('A');
    context
  });

  let result = pipeline.run_detailed(ImmutableContext {
    value: "X".to_string(),
    error: String::new(),
  });
  assert_eq!(
    result.context,
    ImmutableContext {
      value: "XA".to_string(),
      error: "bad".to_string(),
    }
  );
}

#[test]
fn invalid_error_handler_panics_after_all_post_actions() {
  let calls = Arc::new(Mutex::new(Vec::<String>::new()));
  let mut pipeline = Pipeline::new("bad_handler", true);
  pipeline.on_error_handler(|context: String, _error| {
    drop(context);
    panic!("handler failed")
  });
  pipeline.add_action(failing_action);

  let post_one_calls = calls.clone();
  pipeline.add_post_action(move |value: String| {
    post_one_calls.lock().unwrap().push("post1".to_string());
    value
  });
  let post_two_calls = calls.clone();
  pipeline.add_post_action(move |value: String| {
    post_two_calls.lock().unwrap().push("post2".to_string());
    value
  });

  let result = catch_unwind(AssertUnwindSafe(|| pipeline.run("X".to_string())));
  assert!(result.is_err());
  assert_eq!(
    calls.lock().unwrap().clone(),
    vec!["post1".to_string(), "post2".to_string()]
  );
}

#[test]
fn nested_and_concurrent_runs_keep_control_state_isolated() {
  let mut inner_pipeline = Pipeline::new("inner", true);
  inner_pipeline.add_action(append_and_stop);
  inner_pipeline.add_action(|value: String| format!("{value}I2"));
  inner_pipeline.add_post_action(|value: String| format!("{value}IP"));
  let inner = Arc::new(inner_pipeline);

  let inner_for_outer = inner.clone();
  let mut outer = Pipeline::new("outer", true);
  outer.add_action(move |value: String| {
    format!("{value}:{}", inner_for_outer.run("inner".to_string()))
  });
  outer.add_action(|value: String| format!("{value}O2"));
  assert_eq!(outer.run("outer".to_string()), "outer:innerSIPO2");

  let barrier = Arc::new(Barrier::new(2));
  let mut shared_pipeline = Pipeline::new("shared", true);
  let action_barrier = barrier.clone();
  shared_pipeline.add_action(move |value: String| {
    action_barrier.wait();
    if value == "stop" {
      short_circuit();
    }
    format!("{value}A")
  });
  shared_pipeline.add_action(|value: String| format!("{value}B"));
  shared_pipeline.add_post_action(|value: String| format!("{value}P"));
  let shared = Arc::new(shared_pipeline);

  let stop_pipeline = shared.clone();
  let stop_thread = thread::spawn(move || stop_pipeline.run("stop".to_string()));
  let go_pipeline = shared.clone();
  let go_thread = thread::spawn(move || go_pipeline.run("go".to_string()));

  assert_eq!(stop_thread.join().unwrap(), "stopAP");
  assert_eq!(go_thread.join().unwrap(), "goABP");
}

#[test]
fn freeze_and_legacy_control_adapter_use_the_same_runner() {
  let mut pipeline = Pipeline::new("frozen", true);
  pipeline.add_action(|value: String| format!("{value}A"));
  assert_eq!(pipeline.run("X".to_string()), "XA");
  assert!(pipeline.is_frozen());
  let mutation = catch_unwind(AssertUnwindSafe(|| {
    pipeline.add_action(|value: String| format!("{value}B"));
  }));
  assert!(mutation.is_err());
  assert!(catch_unwind(short_circuit).is_err());

  #[allow(deprecated)]
  {
    let mut legacy = Pipeline::new("legacy", true);
    legacy.add_action_control(|value: String, control: &mut ActionControl<String>| {
      control.short_circuit();
      format!("{value}L")
    });
    legacy.add_action(|value: String| format!("{value}B"));
    assert_eq!(legacy.run("X".to_string()), "XL");
  }
}

struct RecordingObserver {
  events: Arc<Mutex<Vec<String>>>,
}

impl PipelineObserver for RecordingObserver {
  fn on_action_started(&self, _pipeline_name: &str, phase: &str, index: usize, name: &str) {
    self.events.lock().unwrap().push(format!("{phase}:{index}:{name}"));
    short_circuit();
  }
}

#[test]
fn observer_failures_and_control_attempts_do_not_change_semantics() {
  let events = Arc::new(Mutex::new(Vec::new()));
  let mut pipeline = Pipeline::new("observed", true);
  pipeline.observer(RecordingObserver {
    events: events.clone(),
  });
  pipeline.add_action(|value: String| format!("{value}A"));
  pipeline.add_action(|value: String| format!("{value}B"));

  assert_eq!(pipeline.run("X".to_string()), "XAB");
  assert_eq!(
    events.lock().unwrap().clone(),
    vec!["actions:0:s0".to_string(), "actions:1:s1".to_string()]
  );
}

#[test]
fn provider_modes_are_eager_fixed_and_round_robin() {
  let created = Arc::new(AtomicUsize::new(0));
  let factory_counter = created.clone();
  let factory = move || {
    let instance_id = factory_counter.fetch_add(1, Ordering::SeqCst) + 1;
    let mut pipeline = Pipeline::new(format!("p{instance_id}"), true);
    pipeline.add_action(move |value: String| format!("{value}{instance_id}"));
    pipeline
  };

  let pooled = PipelineProvider::pooled(factory, 3);
  assert_eq!(created.load(Ordering::SeqCst), 3);
  assert_eq!(pooled.mode(), PipelineProviderMode::Pooled);
  assert_eq!(pooled.instance_count(), 3);
  assert_eq!(
    (0..5).map(|_| pooled.run(String::new())).collect::<Vec<_>>(),
    vec!["1", "2", "3", "1", "2"]
  );

  let singleton: PipelineProvider<String> =
    PipelineProvider::singleton(Pipeline::<String>::new("singleton", true));
  assert!(Arc::ptr_eq(&singleton.get_pipeline(), &singleton.get_pipeline()));

  let per_event = PipelineProvider::new_instance_per_event(|| Pipeline::<String>::new("event", true));
  assert!(!Arc::ptr_eq(&per_event.get_pipeline(), &per_event.get_pipeline()));
}

#[test]
fn router_selects_provider_without_changing_lifecycle() {
  let trade: Arc<PipelineProvider<String>> = Arc::new(PipelineProvider::singleton(
    Pipeline::<String>::new("trade", true),
  ));
  let quote: Arc<PipelineProvider<String>> = Arc::new(PipelineProvider::singleton(
    Pipeline::<String>::new("quote", true),
  ));
  let trade_route = trade.clone();
  let quote_route = quote.clone();
  let router = PipelineRouter::new(move |event: &String| {
    if event == "TRADE" {
      trade_route.clone()
    } else {
      quote_route.clone()
    }
  });

  assert!(Arc::ptr_eq(&router.get_pipeline_provider(&"TRADE".to_string()), &trade));
  assert!(Arc::ptr_eq(&router.get_pipeline_provider(&"QUOTE".to_string()), &quote));
}

#[test]
fn json_loader_supports_canonical_action_arrays() {
  let mut registry: PipelineRegistry<String> = PipelineRegistry::new();
  registry.register_unary("identity", identity_action);

  let json_text = r#"
{
  "pipeline": "t",
  "type": "unary",
  "preActions": [{"$local": "identity"}],
  "actions": [{"$local": "identity"}],
  "postActions": [{"$local": "identity"}]
}
"#;

  let pipeline = PipelineJsonLoader::new()
    .load_str(json_text, &registry)
    .expect("loader failed");
  assert_eq!(pipeline.run("ok".to_string()), "ok");
}

#[test]
fn remote_actions_use_the_ordinary_pipeline_runner() {
  let server = FixtureServer::start();
  let endpoint = server.url("/remote_hello.txt");

  let mut spec = RemoteSpec::new(endpoint.clone());
  spec.method = "GET".to_string();
  assert_eq!(
    http_step(&spec, &"ignored").expect("http_step failed"),
    REMOTE_FIXTURE_BODY.to_string()
  );

  let json_text = format!(
    r#"
{{
  "pipeline": "t",
  "type": "unary",
  "actions": [{{"$remote": {{"endpoint": "{endpoint}", "method": "GET"}}}}]
}}
"#
  );
  let pipeline = PipelineJsonLoader::new()
    .load_str(&json_text, &PipelineRegistry::new())
    .expect("loader failed");
  assert_eq!(pipeline.run("ignored".to_string()), REMOTE_FIXTURE_BODY);
}

fn handle_fixture_connection(mut stream: TcpStream) {
  let mut request_buffer = [0_u8; 2048];
  let bytes_read = stream.read(&mut request_buffer).unwrap_or(0);
  let request_text = String::from_utf8_lossy(&request_buffer[..bytes_read]);
  let request_line = request_text.lines().next().unwrap_or("");
  let mut parts = request_line.split_whitespace();
  let method = parts.next().unwrap_or("");
  let path = parts.next().unwrap_or("");
  let found = method == "GET" && path == "/remote_hello.txt";
  let response_body = if found { REMOTE_FIXTURE_BODY } else { "not found" };
  let status_line = if found { "HTTP/1.1 200 OK" } else { "HTTP/1.1 404 Not Found" };
  let response_text = format!(
    "{status_line}\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{response_body}",
    response_body.len()
  );
  let _ = stream.write_all(response_text.as_bytes());
}

struct FixtureServer {
  address: String,
  shutdown_sender: mpsc::Sender<()>,
  join_handle: Option<thread::JoinHandle<()>>,
}

impl FixtureServer {
  fn start() -> Self {
    let listener = TcpListener::bind("127.0.0.1:0").expect("bind failed");
    listener.set_nonblocking(true).expect("nonblocking failed");
    let address = listener.local_addr().expect("local_addr failed");
    let (shutdown_sender, shutdown_receiver) = mpsc::channel::<()>();
    let join_handle = thread::spawn(move || loop {
      if shutdown_receiver.try_recv().is_ok() {
        break;
      }
      match listener.accept() {
        Ok((stream, _)) => handle_fixture_connection(stream),
        Err(error) if error.kind() == std::io::ErrorKind::WouldBlock => {
          thread::sleep(Duration::from_millis(5));
        }
        Err(_) => break,
      }
    });
    Self {
      address: format!("http://{address}"),
      shutdown_sender,
      join_handle: Some(join_handle),
    }
  }

  fn url(&self, path: &str) -> String {
    format!("{}{}", self.address, path)
  }
}

impl Drop for FixtureServer {
  fn drop(&mut self) {
    let _ = self.shutdown_sender.send(());
    if let Some(handle) = self.join_handle.take() {
      let _ = handle.join();
    }
  }
}

File: src/typescript/tests/pipeline.test.ts
import assert from "node:assert/strict";
import { createServer, Server } from "node:http";
import { AddressInfo } from "node:net";
import test from "node:test";

import {
  ActionControl,
  InvalidErrorHandlerError,
  Pipeline,
  PipelineJsonLoader,
  PipelineObserver,
  PipelineProvider,
  PipelineProviderMode,
  PipelineRegistry,
  PipelineRouter,
  RemoteSpec,
  http_step,
  shortCircuit,
} from "../src/index.js";

function appendAndStop(value: string): string {
  shortCircuit();
  return value + "S";
}

function fail(value: string): string {
  void value;
  throw new Error("boom");
}

test("run returns context and runDetailed uses the same semantics", async () => {
  const pipeline = new Pipeline<string>("simple")
    .addPreAction((value) => value + "P")
    .addAction((value) => value + "A")
    .addPostAction((value) => value + "Z");

  assert.equal(await pipeline.run("X"), "XPAZ");
  const result = await pipeline.runDetailed("X");
  assert.equal(result.context, "XPAZ");
  assert.equal(result.actionTimings.length, 3);
});

test("shortCircuit preserves phase rules", async () => {
  const main = new Pipeline<string>("main")
    .addAction(appendAndStop)
    .addAction((value) => value + "B")
    .addPostAction((value) => value + "P");
  assert.equal(await main.run("X"), "XSP");

  const pre = new Pipeline<string>("pre")
    .addPreAction(appendAndStop)
    .addPreAction((value) => value + "2")
    .addAction((value) => value + "M")
    .addPostAction((value) => value + "P");
  assert.equal(await pre.run("X"), "XS2P");

  const post = new Pipeline<string>("post")
    .addAction((value) => value + "M")
    .addPostAction(appendAndStop)
    .addPostAction((value) => value + "2");
  const result = await post.runDetailed("X");
  assert.equal(result.context, "XMS2");
  assert.equal(result.shortCircuited, true);
});

test("exception policies capture errors and always run postActions", async () => {
  const continuing = new Pipeline<string>("continue", false)
    .addAction((value) => value + "A")
    .addAction(fail)
    .addAction((value) => value + "B");
  const continued = await continuing.runDetailed("X");
  assert.equal(continued.context, "XAB");
  assert.equal(continued.shortCircuited, false);
  assert.equal(continued.errors[0].actionIndex, 1);

  const stopping = new Pipeline<string>("stop", true)
    .addAction((value) => value + "A")
    .addAction(fail)
    .addAction((value) => value + "B")
    .addPostAction((value) => value + "P");
  const stopped = await stopping.runDetailed("X");
  assert.equal(stopped.context, "XAP");
  assert.equal(stopped.shortCircuited, true);
});

test("invalid error handlers fail after cleanup", async () => {
  const calls: Array<string> = [];
  const pipeline = new Pipeline<string>("bad")
    .onError(() => null as unknown as string)
    .addAction(fail)
    .addPostAction((value) => {
      calls.push("post1");
      return value;
    })
    .addPostAction((value) => {
      calls.push("post2");
      return value;
    });

  await assert.rejects(
    pipeline.run("X"),
    InvalidErrorHandlerError,
  );
  assert.deepEqual(calls, ["post1", "post2"]);
});

test("nested and concurrent async runs keep ambient state isolated", async () => {
  const inner = new Pipeline<string>("inner")
    .addAction(appendAndStop)
    .addAction((value) => value + "I2")
    .addPostAction((value) => value + "IP");
  const outer = new Pipeline<string>("outer")
    .addAction(async (value) => value + ":" + (await inner.run("inner")))
    .addAction((value) => value + "O2");
  assert.equal(await outer.run("outer"), "outer:innerSIPO2");

  let started = 0;
  let release!: () => void;
  const gate = new Promise<void>((resolve) => {
    release = resolve;
  });

  const shared = new Pipeline<string>("shared")
    .addAction(async (value) => {
      started += 1;
      if (started === 2) {
        release();
      }
      await gate;
      if (value === "stop") {
        shortCircuit();
      }
      return value + "A";
    })
    .addAction((value) => value + "B")
    .addPostAction((value) => value + "P");
  const provider = PipelineProvider.singleton(shared);

  const [stopped, continued] = await Promise.all([
    provider.run("stop"),
    provider.run("go"),
  ]);
  assert.equal(stopped, "stopAP");
  assert.equal(continued, "goABP");
});

test("plan freezing and compatibility control use the same runner", async () => {
  const pipeline = new Pipeline<string>("frozen")
    .addAction((value) => value + "A");
  assert.equal(await pipeline.run("X"), "XA");
  assert.equal(pipeline.isFrozen(), true);
  assert.throws(
    () => pipeline.addAction((value) => value + "B"),
    /frozen/,
  );
  assert.throws(shortCircuit, /active Pipeline run/);

  const legacy = new Pipeline<string>("legacy")
    .addAction((value: string, control: ActionControl<string>) => {
      control.short_circuit();
      return value + "L";
    })
    .addAction((value) => value + "B");
  assert.equal(await legacy.run("X"), "XL");
});

test("observers cannot control execution", async () => {
  const events: Array<string> = [];
  const observer: PipelineObserver = {
    onActionStarted: (_pipeline, phase, index, name) => {
      events.push(`${phase}:${index}:${name}`);
      shortCircuit();
    },
  };
  const pipeline = new Pipeline<string>("observed")
    .observer(observer)
    .addAction((value) => value + "A")
    .addAction((value) => value + "B");
  assert.equal(await pipeline.run("X"), "XAB");
  assert.deepEqual(events, ["actions:0:s0", "actions:1:s1"]);
});

test("provider modes are eager, fixed, and round robin", async () => {
  let created = 0;
  const factory = (): Pipeline<string> => {
    created += 1;
    const instanceId = created;
    return new Pipeline<string>(`p${instanceId}`)
      .addAction((value) => value + String(instanceId));
  };

  const pooled = PipelineProvider.pooled(factory, 3);
  assert.equal(created, 3);
  assert.equal(pooled.mode, PipelineProviderMode.POOLED);
  assert.equal(pooled.instanceCount, 3);
  assert.deepEqual(
    await Promise.all([
      pooled.run(""),
      pooled.run(""),
      pooled.run(""),
      pooled.run(""),
      pooled.run(""),
    ]),
    ["1", "2", "3", "1", "2"],
  );

  const singleton = PipelineProvider.singleton(
    new Pipeline<string>("singleton"),
  );
  assert.equal(singleton.getPipeline(), singleton.getPipeline());

  const perEvent = PipelineProvider.newInstancePerEvent(factory);
  assert.notEqual(perEvent.getPipeline(), perEvent.getPipeline());
});

test("PipelineRouter selects providers without changing lifecycle", () => {
  const trade = PipelineProvider.singleton(
    new Pipeline<string>("trade"),
  );
  const quote = PipelineProvider.singleton(
    new Pipeline<string>("quote"),
  );
  const router = new PipelineRouter<string, string>(
    (event) => (event === "TRADE" ? trade : quote),
  );
  assert.equal(router.getPipelineProvider("TRADE"), trade);
  assert.equal(router.getPipelineProvider("QUOTE"), quote);
});

async function identityAction(value: unknown): Promise<unknown> {
  return value;
}

test("JSON loader accepts canonical vocabulary", async () => {
  const registry = new PipelineRegistry();
  registry.register_unary("identity", identityAction);

  const pipeline = new PipelineJsonLoader().load_str(
    `{
      "pipeline": "json",
      "preActions": [{"$local": "identity"}],
      "actions": [{"$local": "identity"}],
      "postActions": [{"$local": "identity"}]
    }`,
    registry,
  );
  assert.equal(await pipeline.run("ok"), "ok");
});

const remoteFixtureBody = "Hello from remote fixture\n";

function fixtureRequestHandler(request: any, response: any): void {
  if (String(request.url ?? "") === "/remote_hello.txt") {
    response.statusCode = 200;
    response.end(remoteFixtureBody);
    return;
  }
  response.statusCode = 404;
  response.end("not found");
}

class FixtureServer {
  private server: Server | null = null;

  async start(): Promise<number> {
    this.server = createServer(fixtureRequestHandler);
    await new Promise<void>((resolve) =>
      this.server!.listen(0, "127.0.0.1", resolve),
    );
    return (this.server.address() as AddressInfo).port;
  }

  async stop(): Promise<void> {
    if (this.server == null) {
      return;
    }
    const server = this.server;
    this.server = null;
    await new Promise<void>((resolve) =>
      server.close(() => resolve()),
    );
  }
}

test("remote adapters execute as ordinary Actions", async () => {
  const server = new FixtureServer();
  const port = await server.start();
  try {
    const spec = new RemoteSpec(
      `http://127.0.0.1:${port}/remote_hello.txt`,
    );
    spec.method = "GET";
    assert.equal(await http_step(spec, "ignored"), remoteFixtureBody);

    const pipeline = new PipelineJsonLoader().load_str(
      `{
        "pipeline": "remote",
        "actions": [{
          "$remote": {
            "endpoint": "http://127.0.0.1:${port}/remote_hello.txt",
            "method": "GET"
          }
        }]
      }`,
      new PipelineRegistry(),
    );
    assert.equal(await pipeline.run("ignored"), remoteFixtureBody);
  } finally {
    await server.stop();
  }
});
